File: pset6/credit/credit.py
import cs50                             

while True:                             # promt the imput as long as the number is not a positive one
    number = cs50.get_int("Number:")
    if number > 0:
        break

digi = len(str(number))                 # get the number of digits
small_num = number / 1000000000000      # helper variable to make the if conditions better readeble

def digitsum(n):                        # function to calculate the sum of digits
    s = 0
    while n:
        s += n % 10
        n //= 10
    return s

def get_checksum(number):               # funktion that returns the checksum
    sum1 = 0
    sum2 = 0
    for i in range(digi, 0, -1):        # loop backwards through the number
        r = number % 10                 # get the last digit
        number //= 10                   # cutoff the last number
        if i % 2 != digi % 2:           # alternate algorithm for every second digit 
            sum1 = sum1 + digitsum(2 * r)
        else:
            sum2 = sum2 + r
    return sum1 + sum2                  # returm the final checksum

if get_checksum(number) % 10 == 0:      # if the checksum ends with a zero it is valide
    if digi == 13 and small_num > 4 and small_num < 5:              # check all possible starting numbers and lengths
        print("VISA")
    elif digi == 16 and 4000 < small_num and small_num < 5000:
        print("VISA")
    elif digi == 15 and 340 < small_num and small_num < 350:
        print("AMEX")
    elif digi == 15 and 370 < small_num and small_num < 380:
        print("AMEX")
    elif digi == 16 and 5100 < small_num and small_num < 5600:
        print("MASTERCARD")
    else:                               # if the card doen't start with known numbers it is considered invalid
        print("INVALID")
else:                                   # if the checksum is not ending in 0 the card is definitely invalid
    print("INVALID")

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>


int get_len(long);                  //function to determine the lenght of the number in digits
int get_checksum(long);             //funcktion to calculate the checksum

int main(void)                      
{
    long number;                    //variable for storing the card number
    
    do
    {
        number = get_long("Number: ");
        //printf("Your number is: %li\n", number);
    }
    while (number < 0);
    
    int digi = get_len(number);     // get the lenght of the input number
    //printf("number has %i digits\n", digi);


    if (get_checksum(number) % 10 == 0)              //if checksum number is correct see if digits and first two numbers match too
    {
        if (digi == 13 && 3999999999999 < number && number < 4999999999999)
        {
            printf("VISA\n");
        }
        else if (digi == 16 && 3999999999999999 < number && number < 4999999999999999)
        {
            printf("VISA\n");
        }
        else if (digi == 15 && 339999999999999 < number && number < 349999999999999)
        {
            printf("AMEX\n");
        }
        else if (digi == 15 && 369999999999999 < number && number < 379999999999999)
        {
            printf("AMEX\n");
        } 
        else if (digi == 16 && 5099999999999999 < number && number < 5600000000000000)
        {
            printf("MASTERCARD\n");
        }                      
        else
        {
            printf("INVALID\n");
        }
    }
    else                                            // if the checksum is not ending in 0 the card is definitely invalid
    {
        printf("INVALID\n");
    }
}

//********************************************************************************************
int get_checksum(long number)
{
    int digi = get_len(number);
    int r;
    int sum = 0;
    int sum2 = 0;
    
    
    for (int i = digi; i > 0; i--)              // start with the last digit which was calculated by get_len()
    {
        r = number % 10;                        // gives the last digit
        //printf("rest %i\n", r);
        number = (number - r) / 10;             // subtract the last digit and divide by 10 
        
        if (i % 2 != digi % 2)                  // sum up digits starting with the second to last
        {
            //printf("2nd r is %i\n", r);
            if (r < 5)                          // numbers smale than 5 are single digit when multiplied by 2
            {
                sum = sum + (2 * r);
            }
            else if (r == 5)                    // numbers 5 and greater need to return the sum of digits 5*2=10 ==> 1+0=1
            {
                sum = sum + 1;
            }
            else if (r == 6)                    // numbers 5 and greater need to return the sum of digits 6*2=12 ==> 1+2=3
            {
                sum = sum + 3;
            }
            else if (r == 7)                    // numbers 5 and greater need to return the sum of digits 7*2=14 ==> 1+4=5
            {
                sum = sum + 5;
            }
            else if (r == 8)
            {
                sum = sum + 7;
            }
            else
            {
                sum = sum + 9;
            }
        }
        else
        {
            sum2 = sum2 + r;                    // sum up every other digit starting with the last one
        }
    }
    //printf("the sum is %i\n", sum);
    //printf("the sum2 is %i\n", sum2);
        
    int checksum = sum + sum2;
    //printf("checksum = %i\n", checksum);
    return checksum;                            // return the value checksum to the function
}

//********************************************************************************************
int get_len(long number)            //function to determine the lenght of the number in digits
{
    int digi = 0;                   // variable for storing the number of digits in the given number
    
    while (number >= pow(10, digi))
    {
        digi++;
    }
    return digi;
}